### PR TITLE
BFD-704: Made BFD Pipeline orchestration more flexible

### PR DESCRIPTION
**Pull request by karlmdavis**
### Change Details

Everything is now explicitly part of a "job", and scheduled as such. This will make it much simpler for new jobs to be added, e.g. those related to DC Geo and Synthea.

Notes:

1. To figure out what all is going on here, I'd suggest looking at `PipelineApplication`, `PipelineManager`, and `PipelineManagerIT`, in that order.
2. This design is intended to enable an easy-ish transition towards running jobs on more than one node, with state synchronized via a DB. The design choices reflect that, though I do think that they're all reasonable choices even if that enhancement never happens.
3. I've left some `TODO`s in the code on purpose. I may or may not address those prior to final approval -- haven't decided if they're important enough yet.
4. I removed `RifLoaderIdleTasks` and friends as it turns out that code was quite broken. It apparently only worked accidentally; it was not working as intended, at all. Since we no longer need those jobs, it wasn't worth trying to refactor and fix all of it.

### Acceptance Validation

1. This is concurrent orchestration code, and as such, inherently complex. However, I tried hard to minimize and manage that complexity as much as possible. Did I succeed well enough?
2. Is the code correct?
3. Is it reasonably obvious how to create a new job and ensure that it gets run?

### Feedback Requested

(see above)

### External References

- [BFD-704](https://jira.cms.gov/browse/BFD-704)

### Security Implications

This PR has no identified security implications.

- [ ] new software dependencies

- [ ] altered security controls

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java
Comment (mackec1) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java:77]: Small nit; personally I like to formalize Singleton class by making a private constructor and using a static public getInstance(). For example:

private static PipelineManager myPipelineManager = null;

public static getInstance(MetricRegistry appMetrics, PipelineJobRecordStore jobRecordStore) {
    if (null == myPipelineManager) {
      myPipelineManager = new PipelineManager(appMetrics, jobRecordStore);
   }
  return myPipelineManager;
}

Never have to worry about someone trying to do something they shouldn't!

Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java:77]: Doesn't that make things more challenging to unit test? I generally have tests that construct things in all sorts of different ways, with various mocked dependencies. Can't do that with a private constructor.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java
Comment (Dranith) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java:242]: Interesting thought: Does it make sense to even have a separate `AtomicInteger` here?  Since you are using a `ConcurrentHashMap` from reading `size()` it seems to have the same guarentees as `AtomicInteger` does and might actually provide more accuracy.

For example in this case, say you perform the `put` and another thread reads `jobRunningCounter`.  That will result in `jobsRunningCounter` being less than the number running but `jobActiveHandles.size()` will be correct.  

Now you have a written contract above in documentation that says this can only be incremented or read from one thread, so this works .. but it might still be simpler to just use `.size()`

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java
Comment (Dranith) [inline apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java:146]: Agree on the TODO, but how would you know where you stopped / are picking up from?  Are RIF file loads idempotent?

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java
Comment (Dranith) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java:134]: Just thinking out loud, but this is the sort of thing that the Java Service Provider Interfaces are designed for.  That would allow new jobs to be discovered and picked up by just dropping them on the classpath for the `PipelineApplication`.

The code here definitely a more explicit approach, and I can see the value of that though but it feels like a "plugin" architecture could make sense here.

**Review by Dranith — APPROVED** [review]
Looks good.  I just really had a couple thoughts/questions about certain parts.  Nothing major though.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java:134]: I think heading down the SPI/plugin route would _also_ require a dependency injection provider, e.g. Spring Beans or CDI or Guice, right? I dunno'; I should be in bed, so perhaps I'm wrong.

If so, though, I've very intentionally avoided that here. I used to be a big Guice/Dagger stan but these days find that they add way more complexity than they remove.

Even aside from that, given the small number of jobs we have (and are likely to have for the next couple years), I think I prefer how easy it is to figure out from what's going on just by looking at `PipelineApplication`. That would definitely need to be re-evaluated if the number of jobs grown past some point, e.g. towards 10.

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/CcwRifLoadJob.java:146]: Very idempotent.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java:242]: That's potentially a good catch. I added `jobActiveHandles` very late in development and didn't consider whether or not it could replace `jobsRunningCounter`. I'll ponder that a bit.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java:100]: Actually, it was a bit incorrect: `s/SchedulerJob/VolunteerJob`.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java:58]: I think we'll eventually hit a point where we want this to be dynamic, but right now, we only have the one "real" job (for the CCW load), plus `VolunteerJob` and `CronTriggerJob`. I'd hold off on making it dynamic until that count goes up a bit.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/DefaultDataSetMonitorListener.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/DefaultDataSetMonitorListener.java:56]: I honestly can't recall if S3 offers that particular guarantee explicitly, but at a minimum, it's Java API must, as we haven't had any errors related to that in... 3+ years.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java:126]: Yeah, I definitely ought to `s/must/should` on that one. More of an architecture/design note.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/AppConfiguration.java
Comment (brianburton) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/AppConfiguration.java:288]: Loading all of the configuration variables here creates a very tight dependency between the app package and the job package implementations.  It would be nice if we could delegate configuration loading to the individual jobs rather than centralize it here.  Perhaps by passing a BiFunction<String,String,String> or some equivalent interface of our own to their constructor to abstract config setting lookup so we could choose between environment variables or java properties later without changing the submodule configs.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java
Comment (brianburton) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java:142]: jobRegistry needs to be ConcurrentHashMap since we are using stream() in various methods.  Raw HashMap will throw ConcurrentModificationException if anything is added to or removed from the map during iteration/streaming over it.

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/jobs/store/PipelineJobRecordStore.java
Comment (brianburton) [inline apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/jobs/store/PipelineJobRecordStore.java:34]: jobRecords needs to be ConcurrentHashMap since we are using stream() in various methods.  Raw HashMap will throw ConcurrentModificationException if anything is added to or removed from the map during iteration/streaming over it.

**Review by brianburton — COMMENTED** [review]
Looks good.  Much more flexible than the old implementation and should work for DC-Geo's needs. I would like to see the ConcurrentModificationException issues addressed though.

**Review by mackec1 — APPROVED** [review]
Ran IT stuff under debugger to see/understand the underpinnings in action. Lots of good work here.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java
Comment (Dranith) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java:134]: No Java SPI doesn't require CDI.  It is in the base spec.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java]: Yup: it was safe to do so! And in looking through things to make that change, I caught a race condition in `PipelineManager.stop()` that I'd missed, so thanks!

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/AppConfiguration.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/AppConfiguration.java:288]: Yeah, I was thinking about that myself; it's a good insight. I opted to not address it for now, but I expect it'll be one of those things we come back to in a future PR.

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/jobs/store/PipelineJobRecordStore.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/jobs/store/PipelineJobRecordStore.java:34]: My brain's a bit fried after the morning's work, so I'll need a break before I try to reason through this all the way, but at first glance, I think you're right. Thanks!

Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/jobs/store/PipelineJobRecordStore.java]: Yup: you were correct. Fixed. Thanks!

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineApplication.java:134]: I'm going to defer this, at least for now. At the moment, I'm more in favor of less magic (heh, wordplay). It's a good suggestion, though, and one worth considering in the future, particularly if we end up with large numbers of jobs.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java:146]: I agree that's the case now and likely to remain the case in the future.  But someone less familiar with the code and unaware of the distinction could come along later and add a call to registerJob() or make some other change without knowing the risk.

**Review by brianburton — APPROVED** [review]
My comments from previous review have been addressed.  Looks like a good foundation for pipeline job management going forward.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/PipelineManager.java:146]: I understand the concern, but I think it's low risk, just due to how the application works: hard to see why someone would lazy-/late-register jobs, rather than doing so in `PipelineApplication`.

Regardless, though, if concurrency is a concern, then just making the collection concurrent likely wouldn't be enough. You can end up with all sorts of goofy race conditions, even with a "thread safe" collection (as evidenced by the recent commits here). I think, if we want to address it, the right fix is to ensure that the registration API isn't visble to anything other than `PipelineApplication` -- probably by moving jobs into the constructor. Which... is a good idea, actually, and something we should consider doing in the future. This whole thing just went through so many iterations that there are still some goofy bits like that hanging around.

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/volunteer/VolunteerJob.java
Comment (mackec1) [inline apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/volunteer/VolunteerJob.java:82]: Just curious how much thrashing might occur here; if <= 0 it immediately goes back to loop.

File: apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/jobs/store/PipelineJobRecord.java
Comment (mackec1) [inline apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/jobs/store/PipelineJobRecord.java:51]: Love the comment!

File: apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/volunteer/VolunteerJob.java
Comment (karlmdavis) [reply apps/bfd-pipeline/bfd-pipeline-app/src/main/java/gov/cms/bfd/pipeline/app/volunteer/VolunteerJob.java:82]: What do you mean by thrashing? It will definitely spend a lot of time waiting for something to do, but that's on purpose. If we ever move to running jobs across multiple nodes, oversubscribing becomes bad, as you could have a busy node "squatting" on jobs that it doesn't have capacity to work (and not letting any other nodes work them, either).

**Review by karlmdavis — COMMENTED** [review]
Saving notes from review meeting.

**Review by rtravitz — APPROVED** [review]
Not much additional to add at this point, beyond questions asked/addressed in the review meeting. After pulling it down and tracing through, the architecture makes sense to me, and it's probably time to get it out there and battle-tested.

**Review by JeremyIR — APPROVED** [review]
Echoing what Ryan said with regards to last meeting. Looks good.

**Review by rtravitz — DISMISSED** [review]
Re-approving after fixed conflicts.

**Review by brianburton — APPROVED** [review]
Looks good!

**Review by rtravitz — APPROVED** [review]

**Pull request merged**
